mythril_core/src/ioapic.rs

```diff
@@ -0,0 +1,677 @@
+    /// for the given ID.
+    ///
+    /// While reading from the IO Redirect Table is reasonably simple,
+    /// when writing to the table, there are several read-only
+    /// bits we should not attempt to write to.
+    ///
+    /// See section 3.2.4 of the I/O APIC specification.
+    pub fn write_ioredtbl_raw(&self, id: u8, val: u64) {
+        let base_reg = 0x10 + (id * 2);
+        let low_order = (val & 0xffffffff) as u32;
+        let high_order = (val >> 32) as u32;
+        unsafe {
+            self.write_raw(base_reg, low_order);
+            self.write_raw(base_reg + 1, high_order);
+        }
+    }
+
+    /// Write to the IO Redirect Table Register for a given id.
+    pub fn write_ioredtbl(&self, id: u8, entry: IoRedTblEntry) -> Result<()> {
+        let val: u64 = entry.into();
+        if id > 23 {
+            Err(Error::InvalidValue(format!(
+                "I/O APIC IO Redirect Table register`0x{:x}` too large",
+                id
+            )))
+        } else if (val & !IOREDTBL_RW_MASK) != 0 {
+            Err(Error::InvalidValue(format!(
+                "Read-only IO Redirect Table Entry bits set: 0x{:x}",
+                val & !IOREDTBL_RW_MASK
+            )))
+        } else {
+            self.write_ioredtbl_raw(id, val);
+            Ok(())
+        }
+    }
+}
+
+impl TryFrom<Ics> for IoApic {
+    type Error = Error;
+
+    fn try_from(value: Ics) -> Result<IoApic> {
+        match value {
+            Ics::IoApic {
+                ioapic_addr,
+                gsi_base,
+                ..
+            } => IoApic::new(ioapic_addr, gsi_base),
+            _ => Err(Error::InvalidValue(format!(
+                "Attempting to create an IoApic from: {:?}",
+                value.ics_type()
+            ))),
+        }
+    }
+}
+
+impl fmt::Debug for IoApic {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "IOAPIC: id: {} version: 0x{:x} address: {:p} GSI: {}",
+            self.id(),
+            self.version(),
+            *self.addr_lock.lock(),
+            self.gsi_base
+        )
+    }
+}
+
+/// Type of the I/O APIC Destination Field.
+///
+///  - Physical: bits `59:56` contain an APIC ID
+///  - Logical: bits `63:56` specify the logical address, which
+///    may be a set of processors.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DestinationMode {
+    /// Indicates that the destination is an APIC ID.
+    Physical = 0x00,
+    /// Indicates that the destination is a logical address.
+    Logical = 0x01,
+}
+
+impl TryFrom<u8> for DestinationMode {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<DestinationMode> {
+        match value {
+            0x00 => Ok(DestinationMode::Physical),
+            0x01 => Ok(DestinationMode::Logical),
+            _ => Err(Error::InvalidValue(format!(
+                "Invalid destination mode: 0x{:x}",
+                value
+            ))),
+        }
+    }
+}
```

Comment:
Can we not just use the try_from_primitive stuff we use elsewhere instead of this?

Reply:
For pretty much all of these TryFrom's actually.

Reply:
We can use `try_from_primitive`, but we'll end up with a similar amount of code in the function where this is used with `.map_err` right?

Reply:
Posted https://github.com/mythril-hypervisor/mythril/issues/64 after this is implemented I'll use `TryFromPrimitive` here.